src/main/java/org/folio/consortia/service/impl/PublicationServiceImpl.java

```diff
@@ -0,0 +1,217 @@
+package org.folio.consortia.service.impl;
+
+import static org.folio.consortia.exception.PublicationException.PRIMARY_AFFILIATION_NOT_EXISTS;
+import static org.folio.consortia.exception.PublicationException.TENANT_LIST_EMPTY;
+import static org.folio.consortia.utils.TenantContextUtils.prepareContextForTenant;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Semaphore;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.http.HttpException;
+import org.folio.consortia.domain.dto.PublicationRequest;
+import org.folio.consortia.domain.dto.PublicationResponse;
+import org.folio.consortia.domain.dto.PublicationStatus;
+import org.folio.consortia.domain.entity.PublicationStatusEntity;
+import org.folio.consortia.domain.entity.PublicationTenantRequestEntity;
+import org.folio.consortia.exception.PublicationException;
+import org.folio.consortia.repository.PublicationStatusRepository;
+import org.folio.consortia.repository.PublicationTenantRequestRepository;
+import org.folio.consortia.service.HttpRequestService;
+import org.folio.consortia.service.PublicationService;
+import org.folio.consortia.service.TenantService;
+import org.folio.consortia.service.UserTenantService;
+import org.folio.spring.FolioExecutionContext;
+import org.folio.spring.FolioModuleMetadata;
+import org.folio.spring.scope.FolioExecutionContextSetter;
+import org.springframework.core.task.TaskExecutor;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import lombok.AllArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+
+@Service
+@Log4j2
+@AllArgsConstructor
+public class PublicationServiceImpl implements PublicationService {
+  private final TenantService tenantService;
+  private final UserTenantService userTenantService;
+  private final FolioExecutionContext folioExecutionContext;
+  private final FolioModuleMetadata folioModuleMetadata;
+  private final HttpRequestService httpRequestService;
+  private final TaskExecutor asyncTaskExecutor;
+
+  private final PublicationStatusRepository publicationStatusRepository;
+  private final PublicationTenantRequestRepository publicationTenantRequestRepository;
+  private final ObjectMapper objectMapper;
+  private static final int MAX_ACTIVE_THREADS = 5;
+
+  @Override
+  @SneakyThrows
+  public PublicationResponse publishRequest(UUID consortiumId, PublicationRequest publicationRequest) {
+    FolioExecutionContext centralTenantContext = (FolioExecutionContext) folioExecutionContext.getInstance();
```

Comment:
can't we use folioExecutionContext, without getInstance()?

Reply:
yes, we can!